llvm/tools/sycl-post-link/sycl-post-link.cpp

```diff
@@ -972,6 +1016,11 @@ processInputModule(std::unique_ptr<Module> M) {
     DUMP_ENTRY_POINTS(MDesc.entries(), MDesc.Name.c_str(), 1);
 
     MDesc.fixupLinkageOfDirectInvokeSimdTargets();
+    SmallVector<module_split::ModuleDesc, 2> MMsWithDefaultSpecConsts;
+    if (GenerateDeviceImageWithDefaultSpecConsts)
+      MMsWithDefaultSpecConsts = generateModuleDescWithDefaultSpecConstants(
+          MDesc, Modified, SplitOccurred);
```

Comment:
Emitting that new special device image should be cheaper than handling ESIMD. Therefore, this step should happen _after_ ESIMD handling to avoid doing ESIMD handling for an extra module once again.

Moreover, we likely have a bug in `handleESIMD`:

https://github.com/intel/llvm/blob/c4c9bdbd4efc24b3259025fde8ec9e9d0611f7ed/llvm/tools/sycl-post-link/sycl-post-link.cpp#L875-L880

Spec constants are handled _before_ we may do a merge of two device images. If both ESIMD and non-ESIMD device images contain spec constants, then merging them _after_ handling spec constants would produce an invalid module. Metadata about default values would be incorrectly merged, different spec constants would have assigned the same IDs, etc.

I would suggest the following code structure:

```c++
while (Splitter->hasMoreSplits()) {
  do_split();
  MMs = handleESIMD();
  for (int I = 0, E = MMs.size(); I < E; ++I) { // we only want to process original content of MMs
    processSpecConstants(MMs, I);
  }

  for (M : MMs) {
    saveModule(M)
  }
}

processSpecConstants(MMs, Ind) {
  if (!LowerSpecConstants)
    return;
  auto &M = MMs[Ind];
  if (!M->hasSpecConstants())
    return;
  if (GenerateExtraImage) {
    M2 = Clone(M);
    SpecConstantsPass(M2, mode=hardcode_defaults);
    M2.SpecialFlag = true;
    MMs.push_back(M2);
  }
  SpecConstantsPass(M, mode=mode_from_arguments);
}
```